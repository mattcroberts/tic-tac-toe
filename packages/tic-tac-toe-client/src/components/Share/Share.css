.shareIconContainer {
    display: flex;
    flex-direction: row;
    height: 30px;
    margin-left: 6px;
    line-height: 30px;
}

.shareIcon {
    max-height: 30px;
    width: 30px;
    margin-left: 5px;
}

.shareButton {
    border: 0;
    background: none;
    padding: 0;
}

.inviteButton {
    composes: button from "../../pages/Home/Home.css";
}
