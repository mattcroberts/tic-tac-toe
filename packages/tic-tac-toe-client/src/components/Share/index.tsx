import * as React from "react";
import querystring from "querystring";
import Clipboard from "react-clipboard.js";
import Button, { Size } from "../Button";
import style from "./Share.css";

import copyToClipboard from "../../assets/icons/001-paper.svg";
import facebook from "../../assets/icons/002-facebook.svg";
import messenger from "../../assets/icons/003-messenger.svg";
import whatsapp from "../../assets/icons/004-whatsapp.svg";
import email from "../../assets/icons/005-message.svg";

interface IProps {
    title: string;
    url: string;
}

interface IState {
    active: boolean;
    native: boolean;
}

interface IHrefFuncArgs {
    shareUrl: string;
    text: string;
}

const links = {
    facebook: {
        icon: facebook,
        href: ({ shareUrl, text }: IHrefFuncArgs) =>
            `https://facebook.com/sharer/sharer.php?${new URLSearchParams({
                u: shareUrl,
                quote: text
            }).toString()}`
    },
    whatsapp: {
        icon: whatsapp,
        href: ({ shareUrl, text }: IHrefFuncArgs) =>
            `https://api.whatsapp.com/send?${querystring.stringify({
                text: text + "\n\n" + shareUrl
            })}`
    },
    messenger: {
        icon: messenger,
        href: ({ shareUrl, text }: IHrefFuncArgs) =>
            `https://facebook.com/dialog/send?${new URLSearchParams({
                link: shareUrl,
                app_id: 576699156169689,
                redirect_uri: shareUrl
            }).toString()}`
    },
    email: {
        icon: email,
        href: ({ shareUrl, text }: IHrefFuncArgs) =>
            `mailto:?${querystring.stringify({
                subject: text,
                body: text + "\n\n" + shareUrl
            })}`
    },
    copyToClipboard: {
        icon: copyToClipboard
    }
};

export default class Share extends React.Component<IProps, IState> {
    public state = {
        active: false,
        native: !!navigator.share
    };

    constructor(props: IProps) {
        super(props);
        this.toggleActive = this.toggleActive.bind(this);
    }

    public render() {
        if (this.state.active) {
            return this.state.native
                ? this.renderNativeShare()
                : this.renderFallbackShare();
        } else {
            return (
                <Button
                    size={Size.SMALL}
                    className={style.inviteButton}
                    onClick={this.toggleActive}
                >
                    Invite
                </Button>
            );
        }
    }

    private renderNativeShare() {
        navigator.share({
            title: this.props.title,
            url: this.props.url
        });

        return null;
    }

    private renderFallbackShare() {
        return (
            <div className={style.shareIconContainer}>
                <span>Invite another player with</span>
                {Object.entries(links).map(([key, link]) => {
                    if (key === "copyToClipboard") {
                        return (
                            <Clipboard
                                key={key}
                                component="a"
                                button-href="#"
                                data-clipboard-text={this.props.url}
                            >
                                <img
                                    className={style.shareIcon}
                                    src={link.icon}
                                />
                            </Clipboard>
                        );
                    }

                    return (
                        <a
                            key={key}
                            href={link.href({
                                shareUrl: "https://irix.dev/tictactoe/game/abc",
                                text: this.props.title
                            })}
                            target="_blank"
                        >
                            <img className={style.shareIcon} src={link.icon} />
                        </a>
                    );
                })}
            </div>
        );
    }

    private toggleActive() {
        this.setState(state => ({
            ...state,
            active: !state.active
        }));
    }
}
