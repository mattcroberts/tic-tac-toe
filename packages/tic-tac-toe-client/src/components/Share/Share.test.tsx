import * as React from "react";
import { shallow } from "enzyme";
import url from "url";
import Share from "./index";

describe("Share", () => {
    const defaultProps = {
        title: "test title",
        url: "http://example.com/test/path"
    };

    beforeAll(() => {
        global.open = jest.fn();
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    afterAll(() => {
        delete global.open;
    });

    it("should render", () => {
        const component = shallow(<Share {...defaultProps} />);

        expect(component).toMatchSnapshot();
    });

    it("should show share options on click", () => {
        const component = shallow(<Share {...defaultProps} />);
        component.find("Button").simulate("click");
        expect(component).toMatchSnapshot();

        expect(component.find(".shareButton")).toHaveLength(5);
    });

    it("should share native", () => {
        global.navigator.share = jest.fn();
        const component = shallow(<Share {...defaultProps} />);
        component.setState({ native: true });
        component.find("Button").simulate("click");
        expect(component.find(".shareButton")).toHaveLength(0);

        delete global.navigator.share;
    });

    it("should open window for facebook share", () => {
        const component = shallow(<Share {...defaultProps} />);
        component.find("Button").simulate("click");
        component
            .find(".shareButton")
            .at(0)
            .simulate("click");

        expect(global.open).toHaveBeenCalled();
        const openedPath = url.parse(
            decodeURIComponent(global.open.mock.calls[0][0]),
            true
        );

        expect(global.open.mock.calls).toHaveLength(1);

        expect(openedPath.host).toEqual("facebook.com");
        expect(openedPath.pathname).toEqual("/sharer/sharer.php");
        expect(openedPath.query.u).toEqual("http://example.com/test/path");
        expect(openedPath.query.quote).toEqual("test title");
    });

    it("should open window for whatsapp share", () => {
        const component = shallow(<Share {...defaultProps} />);
        component.find("Button").simulate("click");
        component
            .find(".shareButton")
            .at(1)
            .simulate("click");

        expect(global.open).toHaveBeenCalled();
        const openedPath = url.parse(
            decodeURIComponent(global.open.mock.calls[0][0]),
            true
        );

        expect(global.open.mock.calls).toHaveLength(1);

        expect(openedPath.host).toEqual("api.whatsapp.com");
        expect(openedPath.pathname).toEqual("/send");
        expect(openedPath.query).toEqual({
            text: `test title

http://example.com/test/path`
        });
    });

    it("should open window for messenger share", () => {
        const component = shallow(<Share {...defaultProps} />);
        component.find("Button").simulate("click");
        component
            .find(".shareButton")
            .at(2)
            .simulate("click");

        expect(global.open).toHaveBeenCalled();
        const openedPath = url.parse(
            decodeURIComponent(global.open.mock.calls[0][0]),
            true
        );

        expect(global.open.mock.calls).toHaveLength(1);

        expect(openedPath.host).toEqual("facebook.com");
        expect(openedPath.pathname).toEqual("/dialog/send");
        expect(openedPath.query).toEqual({
            app_id: "576699156169689",
            link: "http://example.com/test/path",
            redirect_uri: "http://example.com/test/path"
        });
    });

    it("should open window for email share", () => {
        const component = shallow(<Share {...defaultProps} />);
        component.find("Button").simulate("click");
        component
            .find(".shareButton")
            .at(3)
            .simulate("click");

        const openedPath = url.parse(global.open.mock.calls[0][0], true);

        expect(global.open).toHaveBeenCalledTimes(1);

        expect(openedPath.protocol).toEqual("mailto:");
        expect(openedPath.query).toEqual({
            subject: "test title",
            body: "test title\n\nhttp://example.com/test/path"
        });
    });
});
