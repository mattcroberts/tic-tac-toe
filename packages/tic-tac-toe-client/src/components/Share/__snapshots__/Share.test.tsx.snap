// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Share should render 1`] = `
<Button
  className="inviteButton"
  onClick={[Function]}
  size="small"
>
  Invite
</Button>
`;

exports[`Share should show share options on click 1`] = `
<div
  className="shareIconContainer"
>
  <span>
    Invite another player with
  </span>
  <button
    className="shareButton"
    key="facebook"
    onClick={[Function]}
  >
    <img
      className="shareIcon"
      src={
        Object {
          "": "",
          "0": "0",
          "00": "00",
        }
      }
    />
  </button>
  <button
    className="shareButton"
    key="whatsapp"
    onClick={[Function]}
  >
    <img
      className="shareIcon"
      src={
        Object {
          "": "",
          "0": "0",
          "00": "00",
        }
      }
    />
  </button>
  <button
    className="shareButton"
    key="messenger"
    onClick={[Function]}
  >
    <img
      className="shareIcon"
      src={
        Object {
          "": "",
          "0": "0",
          "00": "00",
        }
      }
    />
  </button>
  <button
    className="shareButton"
    key="email"
    onClick={[Function]}
  >
    <img
      className="shareIcon"
      src={
        Object {
          "": "",
          "0": "0",
          "00": "00",
        }
      }
    />
  </button>
  <ClipboardButton
    className="shareButton"
    data-clipboard-text="http://example.com/test/path"
    key="copyToClipboard"
    onClick={[Function]}
  >
    <img
      className="shareIcon"
      src={Object {}}
    />
  </ClipboardButton>
</div>
`;
